### PR TITLE
Add methods to ResultSet

### PR DESCRIPTION
**Pull request by romkatsu**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | https://github.com/yiisoft/request-model/issues/16

### CODE REVIEW

File: src/ResultSet.php [hunk 1]
```diff
@@ -48,4 +48,27 @@ public function getIterator(): ArrayIterator
...
+            }
+        }
+
+        return $resultSet;
+    }
+
+    public function isValid(): bool
+    {
+        foreach ($this->results as $result) {
+            if (!$result->isValid()) {
```
Comment (xepozz) [inline src/ResultSet.php:67]: Please use `addResult` method to catching an added new errored result:
If an added result has errors then whole ResultSet will have an error.

File: src/ResultSet.php [hunk 2]
```diff
@@ -48,4 +48,27 @@ public function getIterator(): ArrayIterator
...
+                return false;
+            }
+        }
+
+        return true;
```
Comment (xepozz) [inline src/ResultSet.php:72]:
```suggestion
        return $this->isValid;
```

File: src/ResultSet.php [hunk 3]
```diff
@@ -48,4 +48,27 @@ public function getIterator(): ArrayIterator
...
     }
+
+    public function getErrors(): self
+    {
+        $resultSet = new self();
+        foreach ($this->results as $attribute => $result) {
+            if (!$result->isValid()) {
+                $resultSet->addResult($attribute, $result);
+            }
+        }
```
Comment (xepozz) [inline src/ResultSet.php:59]: Are you sure that you want to get ResultSet instead of an array?

Comment (romkatsu) [reply src/ResultSet.php:64]: It seems logical to me

File: src/ResultSet.php [hunk 4]
```diff
@@ -18,18 +18,23 @@ final class ResultSet implements IteratorAggregate
      * @var Result[]
      */
     private array $results = [];
+    private bool $isValid = true;
+
+    public function addResult(string $attribute, Result $result): void
+    {
+        if (!$result->isValid()) {
+            $this->isValid = false;
```
Comment (xepozz) [inline src/ResultSet.php:26]: Please move it lower to 37 line

File: src/ResultSet.php (hunk 3 above)
Comment (xepozz) [reply src/ResultSet.php:64]: Why not just flatten array? It will be already invalid so you need only attributes and their results.

**Review by xepozz — CHANGES_REQUESTED** [review]

File: src/ResultSet.php (hunk 4 above)
Comment (romkatsu) [reply src/ResultSet.php:26]: This won't work if the first result is error

File: src/ResultSet.php (hunk 3 above)
Comment (romkatsu) [reply src/ResultSet.php]: fixed

**Review by xepozz — APPROVED** [review]

File: src/ResultSet.php [hunk 5]
```diff
@@ -48,4 +53,21 @@ public function getIterator(): ArrayIterator
     {
         return new ArrayIterator($this->results);
     }
+
+    public function getErrors(): array
```
Comment (andrew-demb) [inline src/ResultSet.php:57]: Can you describe in phpdoc `@return` tag explicitly specify what items user can expect to be presented in return result?

**Review by samdark — APPROVED** [review]

File: src/ResultSet.php [hunk 6]
```diff
@@ -48,4 +53,21 @@ public function getIterator(): ArrayIterator
...
+    {
+        $errors = [];
+        foreach ($this->results as $attribute => $result) {
+            if (!$result->isValid()) {
+                $errors[$attribute] = $result;
```
Comment (samdark) [inline src/ResultSet.php:62]: Did you mean to collect errors?

```suggestion
                $errors[$attribute] = $result->getErrors();
```

File: src/ResultSet.php (hunk 5 above)
Comment (romkatsu) [reply src/ResultSet.php:60]: done

File: src/ResultSet.php (hunk 6 above)
Comment (romkatsu) [reply src/ResultSet.php]: done

**Comment by samdark** [comment]
👍

File: src/ResultSet.php (hunk 5 above)
Comment (andrew-demb) [reply src/ResultSet.php:60]: As I can see in test (see below), the type should be an `string[][]` (array of arrays of strings), or for psalm `array<string, list<string>>`, not an `string[]`
```php
$this->assertSame(
            ['attribute1' => ['error1', 'error2']],
            $resultSet->getErrors()
);
```

**Pull request merged**